ntp-proto/src/filter.rs

```diff
@@ -451,6 +444,116 @@ fn find_interval(chime_list: &[CandidateTuple]) -> Option<(NtpDuration, NtpDurat
     None
 }
 
+/// Discard the survivor with maximum selection jitter until a termination condition is met.
+fn cluster_algorithm(candidates: &mut Vec<SurvivorTuple>) {
+    // sort the candidates by increasing lambda_p (the merit factor)
+    candidates.sort_by(|a, b| a.metric.cmp(&b.metric));
+
+    let mut qmax_index = 0;
+
+    let mut min_peer_jitter = 2.0e9;
+    let mut max_selection_jitter = -2.0e9;
+
+    loop {
+        // for each candidate
+        for (index, candidate) in candidates.iter().enumerate() {
+            let p = candidate.peer;
+
+            min_peer_jitter = p.statistics.jitter.min(min_peer_jitter);
+
+            // compute the selection jitter
+            let selection_jitter = candidates
+                .iter()
+                .map(|q| p.statistics.offset - q.peer.statistics.offset)
+                .map(|v| v.to_seconds().powi(2))
+                .sum::<f64>()
+                .sqrt();
+
+            if selection_jitter > max_selection_jitter {
+                qmax_index = index;
+                max_selection_jitter = selection_jitter;
+            }
+        }
+
+        // If the maximum selection jitter is less than the
+        // minimum peer jitter, then tossing out more survivors
+        // will not lower the minimum peer jitter, so we might
+        // as well stop.  To make sure a few survivors are left
+        // for the clustering algorithm to chew on, we also stop
+        // if the number of survivors is less than or equal to
+        // NMIN (3).
+        if max_selection_jitter < min_peer_jitter || candidates.len() <= MIN_CLUSTER_SURVIVORS {
```

Comment:
Since we know the length condition already when starting the current iteration, couldn't we lift that to the outer loop and convert it to a while?

Reply:
turns out that the skeleton is taking a shortcut here. We need to return a value, which means we need (effectively) a do-while loop. So sadly we can't move this condition up

Reply:
we are now. Github had some trouble with updating the file it seems